cloudinit/config/cc_grub_dpkg.py

```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
     idevs_empty = util.get_cfg_option_str(
         mycfg, "grub-pc/install_devices_empty", None)
 
-    if ((os.path.exists("/dev/sda1") and not os.path.exists("/dev/sda")) or
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
+    # get the root disk where the /boot directory resides.
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         status_cb=False, capture=True)[0].strip()
+        # if /dev/disk/by-id/ exists, attempt to determine robust name
+        devlist = ""
+        with suppress(FileNotFoundError):
```

Comment:
What `FileNotFoundError`s are we suppressing here? I'd prefer for us to scope this more narrowly, so it's clear that we won't be suppressing errors we don't mean to.

Reply:
We suppress `FileNotFoundError` since `os.listdir()` throws `FileNotFoundError` if the directory does not exist. We cannot assume that `/dev/disk/by-id` will exist on a system, since most KVM based instances do not have this directory at all. 

I really wanted to avoid suppressing this error too, but it cannot be done.
I wish `os.listdir()` just returned None or a empty iterator instead, but since it throws the error, we need to suppress it.

Reply:
Aha, in that case, what do you think about specifically checking for its existence instead:

```suggestion
        by_id_path = "/dev/disk/by-id"
        if os.path.exists(by_id_path):
```

Reply:
With the change to udevadm we no longer need to walk symlinks.